cal_deltaG/oligoarrayaux-3.8/doc/melt.pl.1.in
.TH MELT.PL 1 "@DATE@" "@PACKAGE_NAME@ @PACKAGE_VERSION@" "User Commands"
.SH NAME
melt.pl \- compute entropy and melting temperature for a two-state model
.SH SYNOPSIS
.B melt.pl
.RI [ OPTION ]...
.I FILE
.RI [ FILE ]
.SH DESCRIPTION
\fBmelt.pl\fR performs two-state folding or hybridization calculations.  If one \fIFILE\fR is given, a minimum energy folding is computed for each sequence in the file using \fBhybrid-ss-min\fR; if two \fIFILE\fRs are given, a minimum energy hybridization is computed for each pair of sequences using \fBhybrid-min\fR.  In either case, the free energy, enthalpy, entropy and melting temperature are written to standard output for each calculation.
.P
For hybridization calculations, the total strand concentration must be specified with the \fB--Ct\fR option.
.SH OPTIONS
.TP
.BR -n ", " --NA =RNA|DNA
sets nucleic acid type to RNA or DNA.  Default is RNA.
.TP
.BR -t ", " --temperature =\fItemp
set temperature for energy minimization step to \fItemp\fR.  Default is 37.
.TP
.BR -N ", " --sodium =\fIvalue
sets Sodium ion concentration to \fIvalue\fR molar.  Default is 1.
.TP
.BR -M ", " --magnesium =\fIvalue
sets Magnesium ion concentration to \fIvalue\fR molar.  Default is 0.
.TP
.BR -C ", " --Ct =\fIvalue
sets total strand concentration to \fIvalue\fR molar.
.TP
.BR -p ", " --polymer
use salt corrections for polymers instead of oligomers (the default).
.TP
.BR -I ", " --noisolate
prohibits all isolated basepairs.  Isolated basepairs are helices of length 1; that is, they do not stack on another basepair on either side.  (See also the \fB--prefilter\fR option below.)
.TP
.BR -m ", " --maxbp =\fIvalue
Bases farther apart than \fIvalue\fR cannot form.  Default is no limit.
.TP
.B --circular
treats sequences as circular rather than linear.  Only valid for folding.
.SH OBSCURE OPTIONS
.TP
.B --allpairs
allows basepairs to form between any two nucleotides.  When \fB--allpairs\fR is not specified, only Watson-Crick and wobble basepairs are allowed.
.TP
.BR --maxloop =\fIsize
sets the maximum size of bulge/interior loops to \fIsize\fR.  Default is 30.
.TP
.B --nodangle
removes single-base stacking from consideration.
.TP
.B --simple
makes the penalty for multibranch loops constant rather than affine.
.TP
.BR --prefilter =\fIvalue1\fR[,\fIvalue2\fR]
Sets the prefilter to filter out all basepairs except those in groups of \fIvalue2\fR adjacent basepairs of which \fIvalue1\fR can form.  \fIvalue2\fR is the same as \fIvalue1\fR if unspecified.  Default is 2 of 2.  (See also the \fB--noisolate\fR option above.)
.SH ENVIRONMENT
.IP UNAFOLDDAT
an alternate location from which to read the energy rules.  The default energy rules can be overridden with files in the current directory or in the directory pointed to by UNAFOLDDAT.  \fBmelt.pl\fR looks for each file first in the current directory, then in the directory specified by UNAFOLDDAT and last in \fI/usr/local/share/unafold\fR (or wherever the energy rules were installed).
.SH REFERENCES
@REFERENCES@
.SH "SEE ALSO"
.BR hybrid-ss-min (1),
.BR hybrid-min (1)
.SH AUTHORS
Nick Markham <markhn@rpi.edu> and Michael Zuker <zukerm@rpi.edu>
.SH COPYRIGHT
Copyright (c) 2006-2009, Rensselaer Polytechnic Institute.
.SH AVAILABILITY
UNAFold is available from http://dinamelt.bioinfo.rpi.edu/unafold/.  Commercial use requires a license; see http://dinamelt.bioinfo.rpi.edu/commercial.php.


cal_deltaG/oligoarrayaux-3.8/scripts/melt.pl
#! /usr/bin/perl -w

use strict;
use warnings;

use Getopt::Long;
Getopt::Long::Configure 'gnu_getopt', 'no_auto_abbrev', 'no_ignore_case';

use constant R => 0.0019872;

sub systemError ($) {
    die ($? == -1 ? "Error: $! from $_[0]\n" : 'Exit status ' . ($? >> 8) . " from $_[0]\n")
}

sub safelyOpen (*$$) {
    open $_[0], $_[1], $_[2] or die "Can't open $_[2]: $!\n";
}
sub version ($) {
    print "$_[0] (OligoArrayAux) 3.8\n";
    print "By Nicholas R. Markham and Michael Zuker\n";
    print "Copyright (C) 2006\n";
    print "Rensselaer Polytechnic Institute\n";
    print "Troy, NY 12810-3590 USA\n";
    exit;
}

sub usage () {
    print <<EOF;
Usage: melt.pl [OPTION]... FILE1 [FILE2]

Options:
-n, --NA=(RNA | DNA) (defaults to RNA)
-t, --temperature=<temperature> (defaults to 37)
-N, --sodium=<[Na+] in M> (defaults to 1)
-M, --magnesium=<[Mg++] in M> (defaults to 0)
-C, --Ct=<total strand concentration>
-p, --polymer
-I, --noisolate
-m, --maxbp=<maximum basepair distance>
    --circular

Obscure options:
    --allpairs
    --maxloop=<maximum bulge/interior loop size> (defaults to 30)
    --nodangle
    --simple
    --prefilter=<filter value>

EOF
    print 'Report bugs to markhn@rpi.edu', "\n";
    exit;
}

my ($temp, $Ct, %options) = (37);
GetOptions \%options, 'version|V' => sub { version('hybrid2.pl') }, 'help|h' => sub { usage() }, 'NA|n=s', 'temperature|t=f' => \$temp, 'maxloop=i', 'allpairs', 'sodium|N=f', 'magnesium|M=f', 'Ct|C=f' => \$Ct, 'polymer|p', 'nodangle', 'simple', 'prefilter=s', 'noisolate|I', 'm|maxbp=i', 'circular' or die $!;

my @args;
foreach (keys %options) {
    if ($_ eq 'allpairs' or $_ eq 'circular' or $_ eq 'nodangle' or $_ eq 'noisolate' or $_ eq 'polymer' or $_ eq 'simple' or $_ eq 'zip') {
	push @args, "--$_";
    } else {
	push @args, "--$_";
	push @args, $options{$_} if defined $options{$_};
    }
}

my $file1 = shift;
unless ($file1) {
    print STDERR "Error: file not specified\nRun 'melt.pl -h' for help\n";
    exit 1;
}
my $file2 = shift;
my $prefix = $file1;
$prefix =~ s/\.seq$//;

if ($file2) {
    unless (defined $Ct) {
	die "Error: two files found but no strand concentration given\n";
    }

    $prefix .= "-$file2";
    $prefix =~ s/\.seq$//;
    system('hybrid-min', -t => $temp, -T => $temp, @args, $file1, $file2) == 0 or systemError('hybrid-min');
    print "dG\tdH\tdS\tTm\n";
    safelyOpen *IN, '<', "$prefix.ct";
    while (<IN>) {
	my ($len, $dG, $dH) = /(\d+)\sdG = ([^\s]+)\sdH = ([^\s]+)/ or next;
	my $dS = ($dH - $dG) / (273.15 + $temp);
	my $homo = not ($len % 2);
	my @bases;
	for (my $i = 1; $i <= $len; ++$i) {
	    my $line = scalar <IN>;
	    my (undef, $base, $prev, $next) = split /\s+/, $line;
	    if (($i == $len / 2 and $next) or ($i == $len / 2 + 1 and $prev)) {
		$homo = 0;
	    }
	    if ($i <= $len / 2) {
		push @bases, $base;
	    } else {
		$homo = 0 unless @bases and $base eq shift @bases;
	    }
	}
	my $factor = $homo ? 1 : 4;
	my $Tm = $dH / ($dS + R * log($Ct / $factor));
	printf "%.1f\t%.1f\t%.1f\t%.1f\n", $dG, $dH, 1000 * $dS, $Tm - 273.15;
    }
    close IN or die $!;
} else {
    my $suffix = 'DH';
    $suffix = 'DHD' if exists $options{NA} and $options{NA} eq 'DNA';
    system('hybrid-ss-min', -t => $temp, -T => $temp, @args, $file1) == 0 or systemError('hybrid-ss-min');
    print "dG\tdH\tdS\tTm\n";
    safelyOpen *DG, '<', "$prefix.dG";
    open DH, '-|', 'ct-energy', -s => $suffix, "$prefix.ct" or die "Can't execute ct-energy: $!";
    scalar <DG>;
    while (defined(my $dG = <DG>) and defined (my $dH = <DH>)) {
	$dG = (split /\s+/, $dG)[1];
	chomp $dH;
	my $dS = ($dH - $dG) / (273.15 + $temp);
	my $Tm = $dH / $dS;
	printf "%.1f\t%.1f\t%.1f\t%.1f\n", $dG, $dH, 1000 * $dS, $Tm - 273.15;
    }
    close DH or die $!;
    close DG or die $!;
}
